core-bundle/src/EventListener/DataContainer/CustomTemplateOptionsListener.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\ContentElement;
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\DataContainer;
+use Contao\Module;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class CustomTemplateOptionsListener
+{
+    /**
+     * @var Controller
+     */
+    private $controller;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var array
+     */
+    private $customTemplates;
+
+    /**
+     * @var string
+     */
+    private $prefix;
+
+    /**
+     * @var string
+     */
+    private $proxyClass;
+
+    public function __construct(ContaoFramework $framework, RequestStack $requestStack, array $customTemplates, string $prefix, string $proxyClass)
+    {
+        /** @var Controller $controller */
+        $controller = $framework->getAdapter(Controller::class);
+
+        $this->controller = $controller;
+        $this->requestStack = $requestStack;
+        $this->prefix = $prefix;
+        $this->customTemplates = $customTemplates;
+        $this->proxyClass = $proxyClass;
+    }
+
+    public function __invoke(DataContainer $dc)
+    {
+        if ($this->isOverrideAll()) {
+            return $this->getOverrideAllTemplates($this->prefix);
+        }
+
+        $defaultTemplate = $this->customTemplates[$dc->activeRecord->type] ?? null;
+
+        // Extract default template from legacy class
+        if (null === $defaultTemplate) {
+            $class = $this->getLegacyClass($dc);
+
+            if (!empty($class) && $class !== $this->proxyClass) {
+                $defaultTemplate = $this->getTemplateFromObject(new $class($dc->activeRecord));
+            }
+        }
+
+        if (null === $defaultTemplate) {
+            $defaultTemplate = $this->prefix.$dc->activeRecord->type;
+        }
+
+        return $this->getTemplateGroup($defaultTemplate);
+    }
+
+    private function getLegacyClass(DataContainer $dc): ?string
+    {
+        switch ($dc->table) {
+            case 'tl_content': return ContentElement::findClass($dc->activeRecord->type);
```

Comment:
Can we use `$this->proxyClass::findClass` here so we don't need to hardcode on the table name?

Reply:
Well, not `$this->proxyClass`, but we could inject another variable into the service for `Contao\ContentElement` and `Contao\Module`.

Reply:
Why? The proxy classes extend from the respective parent, so they inherit that method 😉

Reply:
Indeed 😁